File: app/src/main/java/com/kelsos/mbrc/utilities/RemoteViewIntentBuilder.java
package com.kelsos.mbrc.utilities;

import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.IntDef;
import com.kelsos.mbrc.ui.activities.nav.MainActivity;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

public class RemoteViewIntentBuilder {
  public static final String REMOTE_PLAY_PRESSED = "com.kelsos.mbrc.notification.play";
  public static final String REMOTE_NEXT_PRESSED = "com.kelsos.mbrc.notification.next";
  public static final String REMOTE_CLOSE_PRESSED = "com.kelsos.mbrc.notification.close";
  public static final String REMOTE_PREVIOUS_PRESSED = "com.kelsos.mbrc.notification.previous";
  public static final String CANCELLED_NOTIFICATION = "com.kelsos.mbrc.notification.cancel";
  public static final int OPEN = 0;
  public static final int PLAY = 1;
  public static final int NEXT = 2;
  public static final int CLOSE = 3;
  public static final int PREVIOUS = 4;
  public static final int CANCEL = 5;

  public static PendingIntent getPendingIntent(@ButtonAction int id, Context mContext) {
    switch (id) {
      case OPEN:
        Intent notificationIntent = new Intent(mContext, MainActivity.class);
        return PendingIntent.getActivity(mContext, 0, notificationIntent, PendingIntent.FLAG_UPDATE_CURRENT);
      case PLAY:
        Intent playPressedIntent = new Intent(REMOTE_PLAY_PRESSED);
        return PendingIntent.getBroadcast(mContext, 1, playPressedIntent, PendingIntent.FLAG_UPDATE_CURRENT);
      case NEXT:
        Intent mediaNextButtonIntent = new Intent(REMOTE_NEXT_PRESSED);
        return PendingIntent.getBroadcast(mContext, 2, mediaNextButtonIntent, PendingIntent.FLAG_UPDATE_CURRENT);
      case CLOSE:
        Intent clearNotificationIntent = new Intent(REMOTE_CLOSE_PRESSED);
        return PendingIntent.getBroadcast(mContext, 3, clearNotificationIntent, PendingIntent.FLAG_UPDATE_CURRENT);
      case PREVIOUS:
        Intent mediaPreviousButtonIntent = new Intent(REMOTE_PREVIOUS_PRESSED);
        return PendingIntent.getBroadcast(mContext, 4, mediaPreviousButtonIntent, PendingIntent.FLAG_UPDATE_CURRENT);
      case CANCEL:
        Intent cancelIntent = new Intent(CANCELLED_NOTIFICATION);
        return PendingIntent.getBroadcast(mContext, 4, cancelIntent, PendingIntent.FLAG_UPDATE_CURRENT);
      default:
        throw new IndexOutOfBoundsException();
    }
  }

  @IntDef({
      OPEN,
      PLAY,
      CLOSE,
      PREVIOUS,
      NEXT,
      CANCEL
  }) @Retention(RetentionPolicy.SOURCE) public @interface ButtonAction {
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/utilities/RemoteViewIntentBuilder.kt
package com.kelsos.mbrc.utilities

import android.app.PendingIntent
import android.app.PendingIntent.FLAG_UPDATE_CURRENT
import android.content.Context
import android.content.Intent
import androidx.annotation.IntDef
import com.kelsos.mbrc.ui.navigation.main.MainActivity

object RemoteViewIntentBuilder {
  const val REMOTE_PLAY_PRESSED = "com.kelsos.mbrc.notification.play"
  const val REMOTE_NEXT_PRESSED = "com.kelsos.mbrc.notification.next"
  const val REMOTE_CLOSE_PRESSED = "com.kelsos.mbrc.notification.close"
  const val REMOTE_PREVIOUS_PRESSED = "com.kelsos.mbrc.notification.previous"
  const val CANCELLED_NOTIFICATION = "com.kelsos.mbrc.notification.cancel"
  const val OPEN = 0
  const val PLAY = 1
  const val NEXT = 2
  const val CLOSE = 3
  const val PREVIOUS = 4
  const val CANCEL = 5

  fun getPendingIntent(@ButtonAction id: Int, mContext: Context): PendingIntent {
    when (id) {
      OPEN -> {
        val notificationIntent = Intent(mContext, MainActivity::class.java)
        return PendingIntent.getActivity(
          mContext,
          0,
          notificationIntent,
          FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )
      }
      PLAY -> {
        val playPressedIntent = Intent(REMOTE_PLAY_PRESSED)
        return PendingIntent.getBroadcast(
          mContext,
          1,
          playPressedIntent,
          FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )
      }
      NEXT -> {
        val mediaNextButtonIntent = Intent(REMOTE_NEXT_PRESSED)
        return PendingIntent.getBroadcast(
          mContext,
          2,
          mediaNextButtonIntent,
          FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )
      }
      CLOSE -> {
        val clearNotificationIntent = Intent(REMOTE_CLOSE_PRESSED)
        return PendingIntent.getBroadcast(
          mContext,
          3,
          clearNotificationIntent,
          FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )
      }
      PREVIOUS -> {
        val mediaPreviousButtonIntent = Intent(REMOTE_PREVIOUS_PRESSED)
        return PendingIntent.getBroadcast(
          mContext,
          4,
          mediaPreviousButtonIntent,
          FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )
      }
      CANCEL -> {
        val cancelIntent = Intent(CANCELLED_NOTIFICATION)
        return PendingIntent.getBroadcast(
          mContext,
          4,
          cancelIntent,
          FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )
      }
      else -> throw IndexOutOfBoundsException()
    }
  }

  @IntDef(
    OPEN,
    PLAY,
    CLOSE,
    PREVIOUS,
    NEXT,
    CANCEL
  )
  @kotlin.annotation.Retention(AnnotationRetention.SOURCE)
  annotation class ButtonAction
}
